pkg/machine/compression/sparse_file_writer.go

```diff
@@ -123,7 +123,7 @@ func (sw *sparseWriter) Close() error {
 		return errors.New("file is already closed")
 	}
 	if sw.pendingZeroes != 0 {
-		if holeSize := sw.pendingZeroes - 1; holeSize >= zerosThreshold {
+		if holeSize := sw.pendingZeroes; holeSize >= zerosThreshold {
```

Comment:
AFAICS this will corrupt the image: the code will do a seek(+6 GB) (which does not change the file contents) without writing a trailing non-hole byte; the VM image will be 6 GB shorter.

Reply:
The passing machine test seems to disagree: https://cirrus-ci.com/task/5791451773140992
I am not going to pretend I know how sparse files work in general, I am sure there is good reason for the final write but at least on macos it appears to work like this

Reply:
I meant “corrupt” as in “will have a different length and digest”.

If this is a raw image, I would’t be surprised if the VM using it would (try to?) extend the file by writing to the unallocated parts of the image; i.e. this moves the sparse-file extension / allocation from the decompression time to the VM time.

If this change actually preserves length / digest … I’m extremely confused. (Also, in that case, there should not be any speedup! Creating a 6GB-1 hole and issuing a 1-byte write should not be substantially slower than creating a 6GB hole and not issuing a 1-byte write.)

Reply:
ls -lh did report the full size and du did show the actual size on disk without holes so I guess it worked fine
unless I tested the wrong binary

Feel free test things yourself, I am not saying any of this is correct I am just trying to test things until we can get this insane flake fixed

Reply:
Testing this on Mac with a completely default filesystem, with the `holeSize := sw.pendingZeroes`, `sparse_writer.go` as is otherwise, and
```go

func main() {
	srcBytes := [10_000_000]byte{}
	src := bytes.NewReader(srcBytes[:])
	dest, err := os.OpenFile("t", os.O_RDWR|os.O_CREATE, 0o600)
	if err != nil {
		panic(err)
	}
	sw := NewSparseWriter(dest)

	cnt, err := io.Copy(sw, src)
	if err != nil {
		panic(err)
	}
	fmt.Printf("copy count %d\n", cnt)
	err = sw.Close()
	if err != nil {
		panic(err)
	}
	err = dest.Close()
	if err != nil {
		panic(err)
	}
}
```
I see, per `fs_usage`:
```
open              F=3        (RWC________X)  t
fstat64           F=3
lseek             F=3    O=0x00989680         <SEEK_CUR>
write             F=3    B=0x0
close             F=3  
```
and the created file is 0 bytes long. So the trailing zeroes were not written.

---

With `holeSize := sw.pendingZeroes - 1`, this is seek(N-1)+1-byte write, as expected; but the output is
```console
% ls -la t                   
-rw-------  1 …  …  10000000  … t
% du -h t                    
9,5M	t
% stat -f '%z %b %k' t  
10000000 19536 4096
```
i.e. it does not _look_ sparse.

---

On further testing, a 10M file is not created sparse, but a ~1G one is:
```console
% stat -f '%z %b %k' t
1000000000 8 4096
```

This is going to make experiments so confusing.

The precise trigger seems to be 16 MiB of a hole (16MiB +1 total size), at least in this test case.

Reply:
5 out of 6 successful CI runs https://cirrus-ci.com/task/5005660427911168
The one failure there was not the flake but rather "CI agent stopped responding!" which I assume has nothing to do with code/test changes here and was some infra/cirrus problem

I am not saying my change is correct but it fixes the flake at least.

I take another look tomorrow at your example, what I don't get even if the file will not be sparse the image size is less than 10gb so the local disk is more than fast enough based on my experiment.

Reply:
Yeah, I have _absolutely no idea_ what is going on and if you found a way to avoid the flakes, that’s definitely very valuable.

(I’d probably push for this “don’t write trailing zeroes” mode of `SparseWriter` (assuming that’s what it does) to be opt-in, because not writing the input data as given would be so surprising. But if we know that’s fine for VM images, sure, let’s do that.)

Reply:
if you follow the machine init code there is actually a truncate on the raw disk image based to actual requested VM disk size at the end. So I guess is doesn't matter for our sparse writer to write the remaining zeros because of that. And the truncate doesn't cause the weird write behaviour we are seeing here. So maybe should we try to switch seek over to truncate for the remaining zeros?

Reply:
Sure, why not.

(I’d still like to, eventually, understand what is going on here — the kernel, ~by definition, doesn’t know that the 1-byte write is “the last write”, so I think whatever is happening could also happen in the middle of some other VM image. But that’s not as important as keeping CI sane.)

Reply:
Looks like with truncate is seems to hang even more and I also noticed a hang within the regular write() not at the end in Close() so yes you are right in can happen in the middle...

Reply:
> I’d still like to, eventually, understand what is going on here

There's [an apple developer page on writing sparse files](https://developer.apple.com/documentation/foundation/file_system/about_apple_file_system/#2990507).  IIRC, there are some unique requirements.  The one that struck me was some statement that seemed like it would un-sparse the file if it was re-opened.  Perhaps I'm reading that wrongly.